File: Data_Structures_using_C/LinkedList/Exercise/6.py
class Node:
    def __init__(self, data=None):
        self.data=data
        self.next=None

class CircularLinkedList:
    def __init__(self):
        self.head = None
    def insert(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head=new_node
            new_node.next=self.head
        else:
            current=self.head
            while current.next != self.head:
                current=current.next
            current.next=new_node
            new_node.next=self.head

    def print_list(self):
        if self.head is None:
            print('List is empty')
            return 
        current=self.head
        while True:
            print(current.data)
            current=current.next
            if current == self.head:
                break

def copy_circular_linked_list(original_list):
    if original_list.head is None:
        return None
    new_list=CircularLinkedList()
    current=original_list.head
    while True:
        new_list.insert(current.data)
        current=current.next
        if current == original_list.head:
            break
    return new_list

original_list = CircularLinkedList()
original_list.insert(1)
original_list.insert(2)
original_list.insert(3)
original_list.insert(4)
original_list.insert(5)

print("Original list:")
original_list.print_list()

new_list = copy_circular_linked_list(original_list)

print("New list:")
new_list.print_list()

File: Data_Structures_using_C/LinkedList/Exercise/6.c
#include <stdio.h>
#include <stdlib.h>

// Structure for a node in the circular linked list
struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node with the given data
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the circular linked list
void insertNode(struct Node** head, int data) {
    struct Node* newNode = createNode(data);

    if (*head == NULL) {
        // If the list is empty, make the new node the head
        *head = newNode;
        (*head)->next = *head;
    } else {
        // Traverse to the last node
        struct Node* temp = *head;
        while (temp->next != *head) {
            temp = temp->next;
        }

        // Insert the new node at the end
        temp->next = newNode;
        newNode->next = *head;
    }
}

// Function to display the circular linked list
void displayList(struct Node* head) {
    struct Node* temp = head;

    if (head == NULL) {
        printf("The list is empty.\n");
        return;
    }

    do {
        printf("%d ", temp->data);
        temp = temp->next;
    } while (temp != head);

    printf("\n");
}

// Function to copy a circular linked list
struct Node* copyCircularList(struct Node* original) {
    if (original == NULL) {
        return NULL;
    }

    // Create the head node of the new list
    struct Node* newHead = createNode(original->data);

    // Initialize pointers for traversal
    struct Node* originalCurr = original->next;
    struct Node* newCurr = newHead;

    // Copy the remaining nodes
    while (originalCurr != original) {
        struct Node* newNode = createNode(originalCurr->data);
        newCurr->next = newNode;
        newCurr = newNode;
        originalCurr = originalCurr->next;
    }

    // Connect the last node to the new head to form a circular list
    newCurr->next = newHead;

    return newHead;
}

int main() {
    struct Node* originalList = NULL;
    struct Node* copiedList = NULL;

    // Insert nodes into the original circular linked list
    insertNode(&originalList, 1);
    insertNode(&originalList, 2);
    insertNode(&originalList, 3);
    insertNode(&originalList, 4);
    insertNode(&originalList, 5);

    // Display the original list
    printf("Original List: ");
    displayList(originalList);

    // Copy the original list
    copiedList = copyCircularList(originalList);

    // Display the copied list
    printf("Copied List: ");
    displayList(copiedList);

    return 0;
}
